src/journal/journal-file.c

```diff
@@ -3722,17 +3722,16 @@ void journal_reset_metrics(JournalMetrics *m) {
 void journal_default_metrics(JournalMetrics *m, int fd) {
         char a[FORMAT_BYTES_MAX], b[FORMAT_BYTES_MAX], c[FORMAT_BYTES_MAX], d[FORMAT_BYTES_MAX], e[FORMAT_BYTES_MAX];
         struct statvfs ss;
-        uint64_t fs_size;
+        uint64_t fs_size = 0, fs_avail = 0;
 
         assert(m);
         assert(fd >= 0);
 
-        if (fstatvfs(fd, &ss) >= 0)
+        if (fstatvfs(fd, &ss) >= 0) {
                 fs_size = ss.f_frsize * ss.f_blocks;
-        else {
+                fs_avail = ss.f_bsize * ss.f_bavail;
```

Comment:
hmm, this doesn't look right to me. The determine_path_usage() logic in journald-server already watches f_bavail and is supposed to act on it.

journal_default_metric() is supposed to pick values that a somewhat "stable", i.e. it only judges based on relatively stable "physical" parameters, such as disk size. And then, while we write we look at the actual disk usage and disk free, i.e. more dynamic parameters to test against. With change however, we start testing dynamic values we test frequently against kind-of the same values we test only seldomly. That doesn't appear right to me.

Reply:
OK, so we monitor free disk during runtime, but this doesn't mean much if we pick `keep_free` in a way that prevents us from ever having more than one journal file. What do you think would be the appropriate `keep_free` value in the example given in the commit?

Reply:
maybe just set min_use as a function of the disk size? i.e.:

```c
#define MIN_SIZE_LOW (1*1024*1024)
#define MIN_SIZE_HIGH (16*1024*1024)
m->min_size = CLAMP(fs_size / 50, MIN_SIZE_LOW, MIN_SIZE_HIGH)
```

My thinking here is that the minimum use boundary should probably be smaller on small systems than on big ones, but also not without bounds in either side.

With the above we'd always get some sane limits set, but on beefy machines they'd be a bit beefier than on small emebdded devices. And we'd still never insist on more than 50M...

Does that make sense?

Reply:
I see your point, but I think in this case practicality should trump purity.

> /dev/mapper/live-rw  6.4G  5.7G  648M  91% /

With your approach, fs_size/50 is about 131MB, so we get `min_use==MIN_USE_HIGH==16 MB`. With my patch, we get `min_use` about 65M. 16MB is better than what we have now, but still not the best for the purposes of the reported (who wants to capture debug logs on the live image).

In general, I think tying the usage estimate to free space is much more reasonable than to disk size. The problem with disk size is that it varies very widely (anything from tiny cheap SSDs or memory cards for an rpi, <0.5 GB), to 4TB spinning rust to even higher raid drives. So 2% of the total size can vary by orders of magnitude, and for all bigger sizes we'll always simply get the upper clamp. But it's totally OK to have a huge disk with just a bit of free space remaining that is a tiny fraction of the total disk size, but that is still a relatively large area in absolute terms. E.g. I might have a 4TB spinning rust disk on a server, all filled with videos of kittens, and have 4GB empty for system operation. I think in this case my patch (min_use=.4GB) works much better then 16 MB.

(Also on my machine: I have `/dev/mapper/fedora_krowka-root   78G   69G  5.7G  93% /`, so keep_free is about 11GB, and I'd get almost no logs. That's totally not what I want. I'd be much happier with .57GB of logs as in my patch. In fact I have 4GB, because of the way that we bump min_use to current use.)

Reply:
but still, this would create a weird feedback loop, and it shoudn't be too hard for thinking about cases where this goes awfully wrong, i.e. where the fill level changes drastrically during runtime: you start with an empty disk, install a lot of packages and then suddenly get into ocnflicts with the journal, since it assumes to basically own the whole disk, since it was so empty before.

what about mixing my approach with much lowered keep free limits? i.e. lower the keep free thing to MIN(5%, 100M)?

Reply:
In the OP case:
> /dev/mapper/live-rw 6.4G 5.7G 648M 91% /

keep_free→MIN(327,100)→100 MB.
min_use→16MB.
effective range: 16 MB – 548 MB

In my case:
> /dev/mapper/fedora_krowka-root 78G 69G 5.7G 93% /

keep_free → MIN(4GB, 100MB)→100MB
min_use→16MB
effective range: 16 MB – 5.6 GB
(but then there's the max_use limit, which cuts the range down)

In the hypothetical kitten-video server:
> 4TB 4GB free

keep_free → MIN(209715, 100) → 100 MB
min_use→16MB
effective range: 16 MB – 4.9 GB
(max_use blah blah)

It seems a bit harsh for some cases, but OTOH, the logs don't jump from 0 to the max overnight, they usually grow quickly over weeks or months. OK, I think this should work.